bqskit/ir/gates/parameterized/pauliz.py

```diff
@@ -0,0 +1,103 @@
+"""This module implements the PauliZGate."""
+from __future__ import annotations
+
+from typing import Any
+
+import numpy as np
+import numpy.typing as npt
+import scipy as sp
+
+from bqskit.ir.gates.generalgate import GeneralGate
+from bqskit.ir.gates.qubitgate import QubitGate
+from bqskit.qis.pauliz import PauliZMatrices
+from bqskit.qis.unitary.differentiable import DifferentiableUnitary
+from bqskit.qis.unitary.unitary import RealVector
+from bqskit.qis.unitary.unitarymatrix import UnitaryMatrix
+from bqskit.utils.docs import building_docs
+from bqskit.utils.math import dexpmv
+from bqskit.utils.math import dot_product
+from bqskit.utils.math import pauliz_expansion
+from bqskit.utils.math import unitary_log_no_i
+
+
+class PauliZGate(QubitGate, DifferentiableUnitary, GeneralGate):
+    """
+    A gate representing an arbitrary diagonal rotation.
+
+    This gate is given by:
+
+    .. math::
+
+        \\exp({i(\\vec{\\alpha} \\cdot \\vec{\\sigma_Z^{\\otimes n}})})
+
+    Where :math:`\\vec{\\alpha}` are the gate's parameters,
+    :math:`\\vec{\\sigma}` are the PauliZ Z matrices,
+    and :math:`n` is the number of qubits this gate acts on.
+    """
+
+    def __init__(self, num_qudits: int) -> None:
+        """
+        Create a PauliZGate acting on `num_qudits` qubits.
+
+        Args:
+            num_qudits (int): The number of qudits this gate will act on.
+
+        Raises:
+            ValueError: If `num_qudits` is nonpositive.
+        """
+
+        if num_qudits <= 0:
+            raise ValueError(f'Expected positive integer, got {num_qudits}')
+
+        self._name = f'PauliZGate({num_qudits})'
+        self._num_qudits = num_qudits
+        self.paulis = PauliZMatrices(self.num_qudits)
+        self._num_params = len(self.paulis)
+        if building_docs():
+            self.sigmav: npt.NDArray[Any] = np.array([])
+        else:
+            self.sigmav = (-1j / 2) * self.paulis.numpy
+
+    def get_unitary(self, params: RealVector = []) -> UnitaryMatrix:
+        """Return the unitary for this gate, see :class:`Unitary` for more."""
+        self.check_parameters(params)
+        H = dot_product(params, self.sigmav)
+        eiH = sp.linalg.expm(H)
```

Comment:
If performance is important for your application, there should be a faster way to calculate `eiH` since `H` is diagonal. I would expect the same to be true for the gradient calculation but I would have to dig a bit deeper there. If it is not important, can you put a "TODO" marker so we can come back to this later.

Reply:
I've changed the computation for `get_unitary` but have left a TODO for the gradient part.